feat: implement claiming of prebuilt workspaces

evgeniy-scherbina opened the pull request:
Closes https://github.com/coder/internal/issues/512

evgeniy-scherbina committed:
feat: implement claiming of prebuilt workspaces

evgeniy-scherbina commented on coderd/workspaces.go:
```diff
@@ -635,34 +636,71 @@ func createWorkspace(
+		runningWorkspaceAgentID uuid.UUID
 	)
+
+	prebuilds := (*api.PrebuildsClaimer.Load()).(prebuilds.Claimer)
```
Do we need explicit type-casting here?

dannykopping replied:
Can you elaborate? Would you want to use the atomic pointer directly?

dannykopping replied:
You're right, we don't 👍

evgeniy-scherbina commented on coderd/workspaces.go:
```diff
@@ -635,34 +636,71 @@ func createWorkspace(
+			if err != nil {
+				// TODO: comment about best-effort, workspace can be restarted if this fails...
+				api.Logger.Error(ctx, "failed to retrieve running agents of claimed prebuilt workspace",
+					slog.F("workspace_id", claimedWorkspace.ID), slog.Error(err))
```
Could we just return an error here?

dannykopping replied:
We can; initially I was planning for best-effort but perhaps we want to fail hard here instead now.
We'd just have to ensure the claimed workspace's state is reverted in the DB.

evgeniy-scherbina replied:
not relevant anymore

evgeniy-scherbina commented on coderd/workspaces.go:
```diff
@@ -839,6 +885,32 @@ func requestTemplate(ctx context.Context, rw http.ResponseWriter, req codersdk.C
+
+	// No prebuild available.
+	if claimedID == nil {
+		return nil, nil
```
Returning `nil, nil` is generally bad practice? Return an error instead?

dannykopping replied:
We could, but it's technically an error if no prebuild is available to be claimed.

I don't know if `nil, nil` is _explicitly_ a generally-accepted bad practice, but at least in this case it makes semantic sense.

dannykopping replied:
Looking at this again, I think you're right. We should be returning a sentinel error when there is no prebuild claimable.

evgeniy-scherbina replied:
done: https://github.com/coder/coder/pull/17458/commits/85f79268c998af62d0acfc7953df891b51daa65a

evgeniy-scherbina replied:
@dannykopping 
I noticed I broke tests, I had to add this fix as well: https://github.com/coder/coder/pull/17458/commits/b246589a72804dc8c764059364b1cb89e444977b

We have two different cases when prebuilds are unavailable:
- prebuilds are actually unavailable - DB request returned: `sql.ErrNoRows`
- AGPL license is used

As of now I return same error in both cases:
`ErrNoClaimablePrebuiltWorkspaces = errors.New("no claimable prebuilt workspaces found")`

and I expect this error in `workspaces.go`:
```go
if err != nil && !errors.Is(err, prebuilds.ErrNoClaimablePrebuiltWorkspaces) {
	return xerrors.Errorf("claim prebuild: %w", err)
}
```

Do you want to differentiate these cases and introduce two different errors?
- `ErrNoClaimablePrebuiltWorkspaces`
- `ErrNotEntitledToUsePrebuiltWorkspaces`

dannykopping replied:
Yeah that sounds like a good idea

evgeniy-scherbina replied:
resolved

evgeniy-scherbina commented on coderd/workspaces.go:
```diff
@@ -839,6 +885,32 @@ func requestTemplate(ctx context.Context, rw http.ResponseWriter, req codersdk.C
 }
 
+func claimPrebuild(ctx context.Context, claimer prebuilds.Claimer, db database.Store, logger slog.Logger, req codersdk.CreateWorkspaceRequest, owner workspaceOwner) (*database.Workspace, error) {
+	prebuildsCtx := dbauthz.AsPrebuildsOrchestrator(ctx)
```
Is it okay to elevate permissions here? Shouldn't we just rely on user permissions?

evgeniy-scherbina commented on coderd/workspaces.go:
```diff
@@ -839,6 +885,32 @@ func requestTemplate(ctx context.Context, rw http.ResponseWriter, req codersdk.C
+		logger.Error(ctx, "unable to find claimed workspace by ID", slog.Error(err), slog.F("claimed_prebuild_id", (*claimedID).String()))
+		return nil, xerrors.Errorf("find claimed workspace by ID %q: %w", (*claimedID).String(), err)
+	}
+	return &lookup, err
```
`return &lookup, nil`?

evgeniy-scherbina commented on coderd/wsbuilder/wsbuilder.go:
```diff
@@ -624,6 +641,11 @@ func (b *Builder) findNewBuildParameterValue(name string) *codersdk.WorkspaceBui
 }
 
 func (b *Builder) getLastBuildParameters() ([]database.WorkspaceBuildParameter, error) {
+	// TODO: exclude preset params from this list instead of returning nothing?
```
Self-review: fix it

evgeniy-scherbina replied:
I removed it.

evgeniy-scherbina commented on enterprise/coderd/coderd.go:
```diff
@@ -1128,3 +1155,19 @@ func (api *API) runEntitlementsLoop(ctx context.Context) {
 }
+
+func (api *API) setupPrebuilds(entitled bool) (agplprebuilds.ReconciliationOrchestrator, agplprebuilds.Claimer) {
+	enabled := api.AGPL.Experiments.Enabled(codersdk.ExperimentWorkspacePrebuilds)
```
maybe call it `prebuildsExperimentEnabled` instead of `enabled` to avoid confusion?
Because in upstream code `enabled` means `feature is enabled`, but here it means `experiment is enabled`.

Also maybe rename `entitled` to `featureEnabled`?

dannykopping replied:
No need to ask permission for changes things like this

Whatever is both a) clear and b) consistent will usually be accepted in code-review.

evgeniy-scherbina replied:
not relevant in this PR anymore

evgeniy-scherbina commented on enterprise/coderd/prebuilds/claim.go:
```diff
@@ -0,0 +1,64 @@
+		// err := store.AcquireLock(ctx, database.GenLockID(fmt.Sprintf("prebuild-user-claim-%s", userID.String())))
+		// if err != nil {
+		//	return xerrors.Errorf("acquire claim lock for user %q: %w", userID.String(), err)
+		//}
```
1. looks like this is unnecessary?
2. follow-up question, if we don't need this lock? we probably don't need this transaction which executes single query?

dannykopping replied:
1. I think this can be removed, yes. Initially I was planning on having the reconciliation loop orchestrate the claim, but now since the claim is handled imperatively by the API, only one replica can issue this claim call - by definition.

2. Agreed 👍

evgeniy-scherbina replied:
done

evgeniy-scherbina commented on enterprise/coderd/prebuilds/claim.go:
```diff
@@ -0,0 +1,64 @@
+			// No eligible prebuilds found
+			case errors.Is(err, sql.ErrNoRows):
+				// Exit, this will result in a nil prebuildID being returned, which is fine
+				return nil
```
It will return `(nil, nil)`, I think generally this is discouraged?

dannykopping replied:
Responded in https://github.com/coder/coder/pull/17458/files#r2050948229

dannykopping commented on coderd/provisionerdserver/provisionerdserver.go:
```diff
@@ -2462,10 +2462,18 @@ type TemplateVersionImportJob struct {
+	// reused. Context: the agent token is often used in immutable attributes of workspace resource (e.g. VM/container)
+	// to initialize the agent, so if that value changes it will necessitate a replacement of that resource, thus
+	// obviating the whole point of the prebuild.
+	RunningWorkspaceAgentID uuid.UUID `json:"running_workspace_agent_id"`
```
I believe this change belongs in the agent work that @SasSwart is busy with, not here.

evgeniy-scherbina replied:
got it, make sense

dannykopping commented on coderd/workspaces.go:
```diff
@@ -635,34 +636,71 @@ func createWorkspace(
+			claimedWorkspace *database.Workspace
+		)
+
+		// If a template preset was chosen, try claim a prebuild.
```
Reminder about consistent language here and elsewhere:

```suggestion
		// If a template preset was chosen, try claim a prebuilt workspace.
```

dannykopping commented on coderd/workspaces.go:
```diff
@@ -635,34 +636,71 @@ func createWorkspace(
+				api.Logger.Error(ctx, "failed to retrieve running agents of claimed prebuilt workspace",
+					slog.F("workspace_id", claimedWorkspace.ID), slog.Error(err))
+			}
+			if len(agents) >= 1 {
```
We should remove all agent-related changes.
This will be covered in https://github.com/coder/coder/pull/17475.

evgeniy-scherbina replied:
I see you already did it.

dannykopping commented on codersdk/deployment.go:
```diff
@@ -103,6 +104,7 @@ var FeatureNames = []FeatureName{
 	FeatureControlSharedPorts,
 	FeatureCustomRoles,
 	FeatureMultipleOrganizations,
+	FeatureWorkspacePrebuilds,
```
I don't think any of this should be included in this PR; it's not necessary for the claiming code to be merged - is it?

evgeniy-scherbina replied:
yes

dannykopping reviewed (commented):
First-pass review

dannykopping committed:
Reverting unncessary changes

Signed-off-by: Danny Kopping <dannykopping@gmail.com>

dannykopping commented on enterprise/coderd/prebuilds/claim_test.go:
```diff
@@ -0,0 +1,427 @@
+	return result, err
+}
+
+func TestClaimPrebuild(t *testing.T) {
```
We need one or more error cases here; for example if the claim fails for whatever reason (db unavailable, etc).

evgeniy-scherbina replied:
Added here: https://github.com/coder/coder/pull/17458/commits/fcdbba8ecc1892af0d94951444023b4dfcd491ac
@dannykopping wdyt is it enough?

dannykopping replied:
There's a lot of duplication both between `TestClaimPrebuild`/`TestClaimPrebuild_CheckDifferentErrors` and the also between the two check funcs within `TestClaimPrebuild_CheckDifferentErrors`.

I'd prefer it if you could incorporate the err tests into `TestClaimPrebuild` to keep things concise, but it doesn't need to block this PR right now.

Aside from that, your tests represent the _behaviour_ not the _implementation_, so a name like `ErrNoClaimablePrebuiltWorkspaces is returned` is too tightly coupled with the code. Rather name it like "no claimable prebuilt workspaces results in error" or something like that.

evgeniy-scherbina replied:
okay, I'll try to do it in a follow-up PR. But it will be a lot of control flags in the test. It's usually hard to follow the logic if you have a lot of control flags which significantly change the logic.

For ex.: if flag is true - successful scenario, flag is false is unsuccessful scenario.

But let's see how it will look like in a follow-up PR.

evgeniy-scherbina replied:
addressed here: https://github.com/coder/coder/pull/17567

dannykopping committed:
Merge branch 'main' of github.com:/coder/coder into yevhenii/512-claim-prebuild

dannykopping committed:
Refactoring

Signed-off-by: Danny Kopping <dannykopping@gmail.com>

dannykopping committed:
Removing unnecessary field

Signed-off-by: Danny Kopping <dannykopping@gmail.com>

dannykopping committed:
make fmt

Signed-off-by: Danny Kopping <dannykopping@gmail.com>

evgeniy-scherbina commented on enterprise/coderd/prebuilds/claim_test.go:
```diff
@@ -0,0 +1,367 @@
+		"claiming an eligible prebuild results in error": {
+			expectPrebuildClaimed:  true,
+			markPrebuildsClaimable: true,
+		},
```
These two test cases are identical

dannykopping replied:
Whoops, that's something I forgot to revert; I was adding an error case but rolled it back (incompletely, it appears) because it would've been too big a divergence from your branch.

evgeniy-scherbina replied:
okay, I'll remove it for now

evgeniy-scherbina commented on enterprise/coderd/prebuilds/claim_test.go:
```diff
@@ -0,0 +1,367 @@
+
+			// Then: a prebuild should have been claimed.
+			require.EqualValues(t, spy.claims.Load(), 1)
+			require.NotNil(t, spy.claims.Load())
```
Not necessary? We already checked it before. Probably typo?

dannykopping commented on enterprise/coderd/prebuilds/claim_test.go:
```diff
@@ -0,0 +1,367 @@
+
+					if !tc.markPrebuildsClaimable {
+						continue
+					}
```
No, I think we need to execute line 161 for line 183 to work

evgeniy-scherbina commented on enterprise/coderd/prebuilds/claim_test.go:
```diff
@@ -0,0 +1,367 @@
+
+			require.NotNil(t, spy.claimedWorkspace.Load())
+			claimed := *spy.claimedWorkspace.Load()
+			require.NotEqual(t, claimed, uuid.Nil)
```
should it be, `claimed.ID`?

dannykopping replied:
Good catch 👍

evgeniy-scherbina commented on enterprise/coderd/prebuilds/claim_test.go:
```diff
@@ -0,0 +1,367 @@
+			// Then: the number of running prebuilds has changed since one was claimed.
+			currentPrebuilds, err := spy.GetRunningPrebuiltWorkspaces(ctx)
+			require.NoError(t, err)
+			require.NotEqual(t, len(currentPrebuilds), len(runningPrebuilds))
```
Self-review:

I guess it's better to use:
`require.Equal(t, len(currentPrebuilds), len(runningPrebuilds))-1`

evgeniy-scherbina commented on enterprise/coderd/prebuilds/claim_test.go:
```diff
@@ -0,0 +1,367 @@
+			require.NotEqual(t, len(currentPrebuilds), len(runningPrebuilds))
+
+			// Then: the claimed prebuild is now missing from the running prebuilds set.
+			current, err := spy.GetRunningPrebuiltWorkspaces(ctx)
```
Self-review: unnecessary call? we just called it above.

evgeniy-scherbina committed:
CR's fixes

evgeniy-scherbina committed:
CR's fixes

evgeniy-scherbina committed:
Update coderd/workspaces.go

Co-authored-by: Danny Kopping <danny@coder.com>

evgeniy-scherbina committed:
CR's fixes

evgeniy-scherbina committed:
CR's fixes

evgeniy-scherbina committed:
CR's fixes

evgeniy-scherbina committed:
CR's fixes

evgeniy-scherbina committed:
fix tests by updating noop.go

dannykopping committed:
Merge branch 'main' of github.com:/coder/coder into yevhenii/512-claim-prebuild

johnstcn commented on coderd/prebuilds/api.go:
```diff
@@ -19,9 +26,38 @@ type ReconciliationOrchestrator interface {
+	CalculateActions(ctx context.Context, state PresetSnapshot) (*ReconciliationActions, error)
+}
+
+type Claimer interface {
+	Claim(ctx context.Context, store database.Store, userID uuid.UUID, name string, presetID uuid.UUID) (*uuid.UUID, error)
+	Initiator() uuid.UUID
```
suggestion: we could keep `database.Store` as a field of `Claimer`, one less parameter.

evgeniy-scherbina replied:
@dannykopping I guess it make sense, we do it for `StoreReconciler`. Are you okay with this change?

evgeniy-scherbina replied:
Fixed: https://github.com/coder/coder/pull/17458/commits/e7455db207e10c24012b2d1b658efee38ce6c661

johnstcn commented on coderd/workspaces.go:
```diff
@@ -636,33 +638,58 @@ func createWorkspace(
 		provisionerDaemons []database.GetEligibleProvisionerDaemonsByProvisionerJobIDsRow
 	)
+
+	prebuildsClaimer := *api.PrebuildsClaimer.Load()
```
nit: I think this could be moved to the `var` declaration inside the tx.

evgeniy-scherbina replied:
done

johnstcn commented on coderd/workspaces.go:
```diff
@@ -636,33 +638,58 @@ func createWorkspace(
+		)
+
+		// If a template preset was chosen, try claim a prebuilt workspace.
+		if req.TemplateVersionPresetID != uuid.Nil {
+			// Try and claim an eligible prebuild, if available.
+			claimedWorkspace, err = claimPrebuild(ctx, prebuildsClaimer, db, api.Logger, req, owner)
+			if err != nil && !errors.Is(err, prebuilds.ErrNoClaimablePrebuiltWorkspaces) {
+				return xerrors.Errorf("claim prebuild: %w", err)
+			}
+		}
+
+		// No prebuild found; regular flow.
+		if claimedWorkspace == nil {
+			now := dbtime.Now()
+			// Workspaces are created without any versions.
+			minimumWorkspace, err := db.InsertWorkspace(ctx, database.InsertWorkspaceParams{
+				ID:                uuid.New(),
+				CreatedAt:         now,
+				UpdatedAt:         now,
+				OwnerID:           owner.ID,
+				OrganizationID:    template.OrganizationID,
+				TemplateID:        template.ID,
+				Name:              req.Name,
+				AutostartSchedule: dbAutostartSchedule,
+				NextStartAt:       nextStartAt,
+				Ttl:               dbTTL,
+				// The workspaces page will sort by last used at, and it's useful to
+				// have the newly created workspace at the top of the list!
+				LastUsedAt:       dbtime.Now(),
+				AutomaticUpdates: dbAU,
+			})
+			if err != nil {
+				return xerrors.Errorf("insert workspace: %w", err)
+			}
+			workspaceID = minimumWorkspace.ID
+		} else {
+			// Prebuild found!
+			workspaceID = claimedWorkspace.ID
+			initiatorID = prebuildsClaimer.Initiator()
```
I'd prefer if this logic were moved inside `wsbuilder`. The whole idea of that package is to encapsulate the logic of creating a workspace build.

evgeniy-scherbina replied:
Requires reworking of `wsbuilder` - so we decided to defer for now.

johnstcn commented on coderd/workspaces.go:
```diff
@@ -839,6 +873,27 @@ func requestTemplate(ctx context.Context, rw http.ResponseWriter, req codersdk.C
+func claimPrebuild(ctx context.Context, claimer prebuilds.Claimer, db database.Store, logger slog.Logger, req codersdk.CreateWorkspaceRequest, owner workspaceOwner) (*database.Workspace, error) {
+	prebuildsCtx := dbauthz.AsPrebuildsOrchestrator(ctx)
+
+	// TODO: do we need a timeout here?
+	claimCtx, cancel := context.WithTimeout(prebuildsCtx, time.Second*10)
+	defer cancel()
+
+	claimedID, err := claimer.Claim(claimCtx, db, owner.ID, req.Name, req.TemplateVersionPresetID)
```
`ClaimPrebuiltWorkspace` simply executes an `UPDATE` which should not require a separate timeout.

evgeniy-scherbina replied:
done, now I'm just using `prebuildsCtx` instead of `claimCtx`

evgeniy-scherbina committed:
test: add failure testcase scenario

evgeniy-scherbina committed:
refactor: revert interface changes

evgeniy-scherbina committed:
refactor: simplify signature of claim method

evgeniy-scherbina committed:
refactor: CR's fixes

evgeniy-scherbina committed:
Merge remote-tracking branch 'origin/main' into yevhenii/512-claim-prebuild

evgeniy-scherbina committed:
refactor: CR's fixes

evgeniy-scherbina committed:
refactor: make lint

evgeniy-scherbina committed:
refactor: fix linter

evgeniy-scherbina committed:
refactor: remove PrebuildsOrchestrator context from claim tests

evgeniy-scherbina committed:
refactor: don't fail test in a goroutine

EdwardAngert committed:
docs: fix ssh coder example in testing-templates doc (#17550)

from @NickSquangler and @angrycub on Slack

Co-authored-by: EdwardAngert <17991901+EdwardAngert@users.noreply.github.com>

jaaydenh committed:
fix: add websocket close handling (#17548)

resolves #17508 

Display an error in the UI that the websocket closed if the user is
still interacting with the dynamic parameters form

<img width="795" alt="Screenshot 2025-04-23 at 17 57 25"
src="https://github.com/user-attachments/assets/15362ddb-fe01-462e-8537-a48302c5c621"
/>

ethanndickson committed:
ci: move go install tools to separate action (#17552)

I think using an older version of mockgen on the schmoder CI broke the
workflow, so I'm gonna sync it via this action, like we do with the
other `make build` dependencies.

matifali committed:
chore(dogfood): allow provider minor version updates (#17554)

Aericio committed:
fix(examples/templates/docker-devcontainer): update folder path and provider version constraint (#17553)

Co-authored-by: M Atif Ali <me@matifali.dev>

ibetitsmike committed:
feat: enable masking password inputs instead of blocking echo (#17469)

Closes #17059

matifali committed:
fix(examples/templates/kubernetes-devcontainer): update coder provider (#17555)

matifali committed:
docs: add automatic release calendar updates in docs (#17531)

matifali committed:
chore(dogfood): add windsurf module (#17558)

johnstcn commented on enterprise/coderd/prebuilds/claim_test.go:
```diff
@@ -0,0 +1,567 @@
+			t.Parallel()
+
+			// Setup.
+			ctx := testutil.Context(t, testutil.WaitMedium)
```
This may need to be bumped to `testutil.WaitLong`.

evgeniy-scherbina replied:
Bumped to `testutil.WaitSuperLong` to be consistent with `require.Eventually()`.

dannykopping committed:
Merge branch 'main' of github.com:/coder/coder into yevhenii/512-claim-prebuild

johnstcn commented on enterprise/coderd/prebuilds/claim_test.go:
```diff
@@ -0,0 +1,567 @@
+			require.Equal(t, expectedPrebuildsCount-1, len(currentPrebuilds))
+
+			// Then: the claimed prebuild is now missing from the running prebuilds set.
+			var found bool
+			for _, prebuild := range currentPrebuilds {
+				if prebuild.ID == claimed.ID {
+					found = true
+					break
+				}
+			}
+			require.False(t, found, "claimed prebuild should not still be considered a running prebuild")
```
suggestion: look into `slices.ContainsFunc` https://pkg.go.dev/slices#ContainsFunc

johnstcn commented on enterprise/coderd/prebuilds/claim_test.go:
```diff
@@ -0,0 +1,567 @@
+				t.Logf("found %d running prebuilds so far, want %d", len(runningPrebuilds), expectedPrebuildsCount)
+
+				return len(runningPrebuilds) == expectedPrebuildsCount
+			}, testutil.WaitSuperLong, testutil.IntervalSlow)
```
I'm wary about you having `testutil.WaitSuperLong` here when you're using `testutil.WaitMedium` for the `context.Context` you define above.

`testutil.Eventually` is a context-aware version of `require.Eventually` that I think could be useful here.

johnstcn reviewed (approved):
I'd like to see one test addition in `coderd/workspaces_test.go` exercising the AGPL codepath.

I have some additional suggestions to the tests in the `enterprise/coderd/prebuilds` package, but apart from that I don't need to review again 👍

evgeniy-scherbina committed:
fix: use userCtx instead of prebuildCtx for claiming prebuild

evgeniy-scherbina committed:
refactor: CR's fixes

evgeniy-scherbina committed:
CR's fixes

evgeniy-scherbina commented:
> I'd like to see one test addition in `coderd/workspaces_test.go` exercising the AGPL codepath.
> 
> I have some additional suggestions to the tests in the `enterprise/coderd/prebuilds` package, but apart from that I don't need to review again 👍

I'll add that test in a follow-up PR.
